beancount/parser/lexer.py

```diff
@@ -132,12 +131,7 @@ def STRING(self, string):
         if '\n' in string:
             num_lines = string.count('\n') + 1
             if num_lines > self.long_string_maxlines_default:
-                # This is just a warning; accept the string anyhow.
-                self.errors.append(
-                    LexerError(
-                        self.get_lexer_location(),
-                        "String too long ({} lines); possible error".format(num_lines),
-                        None))
+                raise ValueError("String too long ({} lines)".format(num_lines))
```

Comment:
We cannot raise an exception from the parser. We need to accumulate errors for rendering later.

Reply:
The exception is raised here and catch in the `BUILD_LEX` where it is translated into a error accumulated like before. This was already done for "unexpected" exceptions raised by `lexer.LexBuilder` methods. Now it is done for all exceptions, simplifying error handling a bit and removing the need to be able to access the scanning location from `lexer.LexBuilder` methods, which in turn allows some decoupling of the scanner from the rest of the code. From the outside the code works exactly as before (I didn't have to modify any test). The same decoupling could be achieved  passing filename and lineno to all methods (in a similar way as what I have done for the `grammar.Builder` methods) but I thought is was a overkill.

Reply:
Also, this simplification comes very handy when moving the construction of token semantic values to compiled code.

Reply:
SGTM. We should comment in both lexer.py and on BUILD_LEX() that these are coupled.